src/importsource.cpp

```diff
@@ -83,4 +96,211 @@ ImportSourcePtr ImportSource::clone() const
+}
+
+bool ImportSource::addUnits(const UnitsPtr &units)
+{
+    auto import = std::dynamic_pointer_cast<ImportedEntity>(units);
+    const auto it = std::find_if(mPimpl->mImports.begin(), mPimpl->mImports.end(), [&import](const std::weak_ptr<ImportedEntity> &p) {
+        return p.lock() == import;
+    });
+
+    if (it != mPimpl->mImports.end()) {
+        return false;
+    }
+
+    size_t index = mPimpl->mImports.size();
+    mPimpl->mImports.push_back(import);
+    mPimpl->mUnits.push_back(index);
+    return true;
+}
+
+bool ImportSource::removeUnits(UnitsPtr &units, bool setEmpty)
+{
+    if (units == nullptr) {
+        return false;
+    }
+
+    auto import = std::dynamic_pointer_cast<ImportedEntity>(units);
+    auto it = std::find_if(mPimpl->mImports.begin(), mPimpl->mImports.end(), [&import](std::weak_ptr<ImportedEntity> &p) {
+        return p.lock() == import;
+    });
+
+    if (it == mPimpl->mImports.end()) {
+        return false;
+    }
+
+    mPimpl->removeItem(it);
+
+    if (setEmpty) {
+        ImportSourcePtr empty = nullptr;
+        units->setImportSource(empty);
+    }
+
+    return true;
+}
+
+bool ImportSource::removeUnits(size_t index, bool setEmpty)
+{
+    if (index >= mPimpl->mUnits.size()) {
+        return false;
+    }
+
+    auto u = units(index);
+
+    auto import = std::dynamic_pointer_cast<ImportedEntity>(u);
+    auto it = std::find_if(mPimpl->mImports.begin(), mPimpl->mImports.end(), [&import](std::weak_ptr<ImportedEntity> &p) {
+        return p.lock() == import;
+    });
+
+    mPimpl->removeItem(it);
+
+    if (setEmpty) {
+        ImportSourcePtr empty = nullptr;
+        u->setImportSource(empty);
+    }
+
+    return true;
+}
+
+bool ImportSource::removeAllUnits()
+{
+    bool status = true;
+    for (size_t u = mPimpl->mUnits.size() - 1; u <= 0; --u) {
+        status = status && removeUnits(u);
+    }
+    return status;
+}
+
+void ImportSource::ImportSourceImpl::removeItem(std::vector<ImportedEntityWeakPtr>::iterator &it)
+{
+    // Find current index for the item to remove.
+    auto index = int(std::distance(mImports.begin(), it));
+    auto itU = std::find(mUnits.begin(), mUnits.end(), index);
+    if (itU != mUnits.end()) {
+        mUnits.erase(itU);
+    } else {
+        auto itC = std::find(mComponents.begin(), mComponents.end(), index);
+        if (itC != mComponents.end()) {
+            mComponents.erase(itC);
+        }
+    }
+
+    // Iterate through components index vector and decrement all indices above the importIndex.
+    auto i = int(mComponents.size() - 1);
+    while ((i >= 0) && (int(mComponents[size_t(i)]) > index)) {
+        mComponents[size_t(i)]--;
+        i--;
+    }
+    // Loop through units vector and decrease its indices too.
+    i = int(mUnits.size() - 1);
+    while ((i >= 0) && (int(mUnits[size_t(i)]) > index)) {
+        mUnits[size_t(i)]--;
+        i--;
+    }
+    // Remove the item from the imports.
+    mImports.erase(it);
+}
```

Comment:
Definitely not keen on all that casting, not least since `size()` returns a `size_t`, so to cast it to an `int` is "risky" although we are very unlikely to ever have more than 2,147,483,647 components/units. Anyway, I would rewrite this method to read:
```
void ImportSource::ImportSourceImpl::removeItem(std::vector<ImportedEntityWeakPtr>::iterator &it)
{
    // Find current index for the item to remove.
    auto index = size_t(std::distance(mImports.begin(), it));
    auto itU = std::find(mUnits.begin(), mUnits.end(), index);
    if (itU != mUnits.end()) {
        mUnits.erase(itU);
    } else {
        auto itC = std::find(mComponents.begin(), mComponents.end(), index);
        if (itC != mComponents.end()) {
            mComponents.erase(itC);
        }
    }

    // Iterate through components index vector and decrement all indices above the item's index.
    static const size_t MAX_SIZE_T = std::numeric_limits<size_t>::max();
    for (size_t i = mComponents.size() - 1; (i != MAX_SIZE_T) && (mComponents[i] > index); --i) {
        --mComponents[i];
    }
    // Loop through units vector and decrease its indices too.
    for (size_t i = mUnits.size() - 1; (i != MAX_SIZE_T) && (mUnits[i] > index); --i) {
        --mUnits[i];
    }
    // Remove the item from the imports.
    mImports.erase(it);
}
```

Reply:
a8276f79